### ISSUE TITLE
`deno install xxx.ts` will replace blank spaces in folder names with "%20" which won't run on Windows 10

### ISSUE THREAD

**Issue by Peybro** [denoland/deno#6408]
It can be fixed by replacing "%20" by " " (blank space) in `C:\Users[user].deno\bin\xxx.cmd`. After that it's finally possible to run `xxx` from anywhere.

deno 1.1.1
v8 8.5.104
typescript 3.9.2

**Comment by SyrupThinker** [comment]
Also backslashes are turned into forwardslashes on Linux:
```
: mkdir 'A\B'
: ls
'A\B'/
: touch 'A\B'/test.ts
: deno install 'A\B'/test.ts 
error: Cannot resolve module "file:///home/syrup/downloads/A/B/test.ts"
: file A\\B/test.ts
A\B/test.ts: empty
```

deno 1.1.1
v8 8.5.104
typescript 3.9.2

**Comment by linde12** [comment]
OK i think i've located the problem with these two issues. 

For the backslash issue i see that `Url::parse` [here](https://github.com/denoland/deno/blob/master/core/module_specifier.rs#L71) is being passed an invalid URL containing a backslash. `Url::parse` is then being lenient it seems because it doesn't Err and instead removes the backslash so that e.g. `file:///some/path/A\\B/test.ts` becomes `file:///some/path/A/B/test.ts`.

The reason that there is a \ being passed to `Url::parse` in module_specifier.rs in the first place is because it does _NOT_ get removed from `Url::from_file_path` [here](https://github.com/denoland/deno/blob/master/core/module_specifier.rs#L153)

However `Url::parse` does (correctly) encode a space as `%20`, so we have to decode it later. I'm also getting issues with this on linux with the filename having a space, e.g. `hello world.ts` gives `error: Invalid executable name: hello%20world` - this is because we are not decoding the filename [here](https://github.com/denoland/deno/blob/master/cli/installer.rs#L105) either.

tl;dr
Backslash issue:
`Url::from_file_path` does _NOT_ strip \ and doesn't URL encode it either
`Url::parse` strips \ instead of returning Err. `Url::parse` also doesn't decode %5C

URL issue:
`Url::from_file_path` encodes " " (space) as %20 so we need to decode it in `infer_name_from_url` and also
I imagine @Peybro is having a similar issue but with the path, i haven't looked into it too much as i'm not on Windows.

I could take a look at solving the issue with `infer_name_from_url` and the Windows "path with space" issue but as for the URL issues i imagine we want to fix them in the url crate.

Maybe we could have a workaround until then? What do you think @ry ?

**Comment by linde12** [comment]
I have created this issue for now https://github.com/servo/rust-url/issues/612

**Comment by linde12** [comment]
I haven't had the time to look at this yet, but i imagine we want to use the `percent_encoding` crate (used by url and developed by the same people, its in the same source repo) to decode the name (and path on windows? haven't looked at this yet)

**Comment by linde12** [comment]
Adding `let stem = percent_encoding::percent_decode(&stem.as_bytes()).decode_utf8().ok()?.to_string();` to the `infer_name_from_url` function removes the percent encoding, but we still get an `error: Invalid executable name: hello world` because of the `EXEC_NAME_RE` regex - maybe its by design that executable names shouldn't be allowed to contain spaces?

**Comment by bartlomieju** [comment]
Possibly related to #7613? @solson does your PR take care of this problem?

**Comment by solson** [comment]
My PR has no effect on Windows. Additionally, I think at least parts of this problem affect Linux as well, e.g.:

```console
$ deno --version
deno 1.5.2 (71d7482, release, x86_64-unknown-linux-gnu)
v8 8.7.220.3
typescript 4.0.5

$ deno install "hello world.ts"
error: Invalid executable name: hello%20world
```

(It never even gets around to creating `.deno/bin/hello world`.)

Though it seems like having a space in a folder name works fine with `deno install` on Linux, even before my PR.
